csrc/scheduler/registry_utils.cpp

```diff
@@ -442,60 +442,62 @@ bool reductionInterferingView(
-
-  auto disjoint_rfactor_sets = scheduler_utils::disjointLogicalSets(fusion);
-  auto disjoint_set_information = scheduler_utils::getDisjointLogicalSetsOf(
-      fusion, reduction_reference, disjoint_rfactor_sets);
-
-  // Convert id's in groups to disjoint_set_ids of disjoint_set_information
-  std::vector<std::vector<int64_t>> disjoint_groups;
-
-  for (const auto& group : groups) {
-    std::vector<int64_t> disjoint_id_sets;
-    for (auto id : group) {
-      auto find_it = std::find(
-          reduction_reference->getLogicalDomain().begin(),
-          reduction_reference->getLogicalDomain().end(),
-          id);
-      NVF_ERROR(
-          find_it != reduction_reference->getLogicalDomain().end(),
-          "Issue with view analysis on reduction like schedule, with reference: ",
-          reduction_reference->toString());
-      auto logical_pos = std::distance(
-          reduction_reference->getLogicalDomain().begin(), find_it);
-      NVF_ERROR(
-          logical_pos <
-              (int64_t)disjoint_set_information.disjoint_set_ids.size(),
-          "Error computing disjoint group on the logical domain of ",
-          reduction_reference->toString());
-      disjoint_id_sets.push_back(
-          disjoint_set_information.disjoint_set_ids[logical_pos]);
+  // Check if any view op could merge IDs in different coalesced_ids.
+  // (1) Build the exact graph
+  IdModel id_model(fusion, false, false, false);
+  id_model.buildExactGraph();
+  const ValGraph& graph = id_model.idGraph(IdMappingMode::EXACT);
+  auto disjoint_sets = graph.disjointValSets();
+  for (auto view : ir_utils::getViewOps(fusion)) {
+    auto tv = view->out();
+    std::vector<IterDomain*> ids_to_merge;
+    for (auto expr : StmtSort::getExprsTo(
+             {tv->getLogicalDomain().begin(), tv->getLogicalDomain().end()})) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        disjoint_sets.mapEntries(merge->inner(), merge->outer());
+        disjoint_sets.mapEntries(merge->inner(), merge->out());
+        // process staged split IDs, remove after mapped.
+        for (auto id : ids_to_merge) {
+          disjoint_sets.mapEntries(merge->out(), id);
+        }
+        ids_to_merge.clear();
+      }
+      // For split, directly map the inner and outer IDs leads to false
+      // positive e.g. the frist reshape in group norm, since only merge op can
+      // cause the merge of IDs in different coalesced_ids. However, we need to
+      // keep track of the split IDs and map them with merged IDs when a merge
+      // op is found. Otherwise, we may miss a case like: reshape: [I2, I3] ->
+      // [8, I2/8, I3] -> [8, I2/8*I3], reduction: [I2, rI3]. This view +
+      // reduction is not supported since I2/8*I3 includes both iter and redu
+      // dims. If we don't process split, I2/8 is not mapped with I3.
+      else if (auto split = dynamic_cast<Split*>(expr)) {
+        ids_to_merge.push_back(split->inner());
```

Comment:
Looks like this part is just accumulating all split domains, and they are merged with the domains of the first merge op. However, these split input and output domains are not guaranteed to have dependencies with the merge input and output domains, right? Aren't we making false groupings as well as missing true groupings?

Reply:
You are right. Can we keep track of the split IDs and when a merged id comes from a set of connected splits, map it with all other IDs in the same set?
```
    // keep track of split IDs, each unordered_set in the vector stores
    // connected split IDs. For example, if we have 3 splits:
    // I0 -> I1, I2
    // I3 -> I4, I5
    // I5 -> I6, I7
    // There are 2 groups: [I0, I1, I2] and [I3, I4, I5, I6, I7]
    // when a merge op is found, e.g. I9 = I8 * I7, we need to map
    // I9 with I8 and all the IDs in [I3, I4, I5, I6, I7].
    std::vector<std::unordered_set<IterDomain*>> ids_in_connected_splits;
```

Reply:
I think there are multiple valid ways and I'm not sure what the best way would be. I don't have any particular way in my mind, but what we have is a graph, and what we need to know is if a certain set of graph nodes are disjoint. Traversal routines in val_graph_visitor.h may help.